src/ffi/server.rs

```diff
@@ -0,0 +1,236 @@
+        let http = non_null! { &mut *http ?= hyper_code::HYPERE_INVALID_ARG };
+        http.0.http2_initial_connection_window_size(if window_size == 0 { None } else { Some(window_size) });
+        hyper_code::HYPERE_OK
+    }
+}
+
+ffi_fn! {
+    fn hyper_http_http2_adaptive_window(http: *mut hyper_http, enabled: c_int) -> hyper_code {
+        let http = non_null! { &mut *http ?= hyper_code::HYPERE_INVALID_ARG };
+        http.0.http2_adaptive_window(enabled != 0);
+        hyper_code::HYPERE_OK
+    }
+}
+
+ffi_fn! {
+    fn hyper_http_http2_max_frame_size(http: *mut hyper_http, frame_size: c_uint) -> hyper_code {
+        let http = non_null! { &mut *http ?= hyper_code::HYPERE_INVALID_ARG };
+        http.0.http2_max_frame_size(if frame_size == 0 { None } else { Some(frame_size) });
+        hyper_code::HYPERE_OK
+    }
+}
+
+ffi_fn! {
+    fn hyper_http_http2_max_concurrent_streams(http: *mut hyper_http, max_streams: c_uint) -> hyper_code {
+        let http = non_null! { &mut *http ?= hyper_code::HYPERE_INVALID_ARG };
+        http.0.http2_max_concurrent_streams(if max_streams == 0 { None } else { Some(max_streams) });
+        hyper_code::HYPERE_OK
+    }
+}
+
+ffi_fn! {
+    fn hyper_http_http2_max_send_buf_size(http: *mut hyper_http, max_buf_size: usize) -> hyper_code {
+        let http = non_null! { &mut *http ?= hyper_code::HYPERE_INVALID_ARG };
+        http.0.http2_max_send_buf_size(max_buf_size);
+        hyper_code::HYPERE_OK
+    }
+}
+
+ffi_fn! {
+    fn hyper_http_http2_enable_connect_protocol(http: *mut hyper_http) -> hyper_code {
+        let http = non_null! { &mut *http ?= hyper_code::HYPERE_INVALID_ARG };
+        http.0.http2_enable_connect_protocol();
+        hyper_code::HYPERE_OK
+    }
+}
+
+ffi_fn! {
+    fn hyper_http_max_buf_size(http: *mut hyper_http, max_buf_size: usize) -> hyper_code {
+        let http = non_null! { &mut *http ?= hyper_code::HYPERE_INVALID_ARG };
+        http.0.max_buf_size(max_buf_size);
+        hyper_code::HYPERE_OK
+    }
+}
+
+ffi_fn! {
+    fn hyper_http_pipeline_flush(http: *mut hyper_http, enabled: c_int) -> hyper_code {
+        let http = non_null! { &mut *http ?= hyper_code::HYPERE_INVALID_ARG };
+        http.0.pipeline_flush(enabled != 0);
+        hyper_code::HYPERE_OK
+    }
+}
+
+ffi_fn! {
+    fn hyper_http_serve_connection(http: *mut hyper_http, io: *mut hyper_io, service: *mut hyper_service) -> *mut hyper_serverconn {
+        let http = non_null! { &mut *http ?= ptr::null_mut() };
+        let io = non_null! { Box::from_raw(io) ?= ptr::null_mut() };
+        let service = non_null! { Box::from_raw(service) ?= ptr::null_mut() };
+
+        Box::into_raw(Box::new(hyper_serverconn(http.0.serve_connection(*io, *service))))
+    } ?= ptr::null_mut()
+}
+
+pub struct hyper_service {
+    service_func: hyper_service_callback,
+    userdata: *mut c_void,
+}
+
+type hyper_service_callback =
+    extern "C" fn(*mut c_void, *mut hyper_request, *mut hyper_response) -> ();
```

Comment:
So, it looks like this would require that in C, the callback is synchronous. It must return a response, it can't return a future. I worry that this might make servers in C just do everything in a blocking style, which would be bad for the server.

Would it be hard to have this callback return a `*mut hyper_task`? I haven't thought too hard about it